src/main.rs

```diff
@@ -44,97 +51,227 @@ struct Command {
-}
 
-fn main() -> Result<()> {
-    let command = Command::parse();
-
-    let list = get_paths(&command.matchers)?;
-    let total_size = list.iter().map(|item| item.size).sum();
-    let template = "{bytes:>10!}/{total_bytes:<10!} {bytes_per_sec:>15!} [{bar:40}]";
-    let progress_style = ProgressStyle::with_template(template)?;
-    let progress = ProgressBar::new(total_size).with_style(progress_style);
-    for item in list {
-        let file_type = check_format(&item.path)?;
-        let mut data = Vec::new();
-        match file_type {
-            FileType::Ncm => {
-                let file = File::open(&item.path)?;
-                let mut dump = Ncmdump::from_reader(file)?;
-                let mut buffer = [0; 1024];
-                while let Ok(size) = dump.read(&mut buffer) {
-                    if size == 0 {
-                        break;
+    fn get_output(
+        &self,
+        file_path: &Path,
+        format: &str,
+        output: &Option<String>,
+    ) -> Result<PathBuf> {
+        let parent = match output {
+            None => file_path.parent().ok_or(Error::PathError)?,
+            Some(p) => Path::new(p),
+        };
+        let file_name = file_path.file_stem().ok_or(Error::PathError)?;
+        let path = parent.join(file_name).with_extension(format);
+        Ok(path)
+    }
+
+    fn get_subfile(&self, dir: PathBuf, depth: u8) -> Result<Vec<PathBuf>> {
+        let mut result = Vec::new();
+        if dir.is_dir() {
+            for entry in dir.read_dir()? {
+                let path = entry?.path();
+                if path.is_file() {
+                    result.push(path);
+                } else if path.is_dir() && self.command.recursive {
+                    if depth < MAX_RECURSIVE_DEPEH {
+                        result.extend(self.get_subfile(path, depth + 1)?);
+                    } else {
+                        self.progress
+                            .println("Folder nesting layers are too deep, skipping")?;
                     }
-                    data.write_all(&buffer[..size])?;
-                    progress.inc(size as u64);
                 }
             }
-            FileType::Qmc => {
-                let file = File::open(&item.path)?;
-                let mut dump = QmcDump::from_reader(file)?;
-                let mut buffer = [0; 1024];
-                while let Ok(size) = dump.read(&mut buffer) {
-                    if size == 0 {
-                        break;
+        }
+        Ok(result)
+    }
+
+    fn get_paths(&self) -> Result<Vec<PathBuf>> {
+        let mut paths = Vec::new();
+        for matcher in &self.command.matchers {
+            for entry in glob(matcher.as_str())? {
+                match entry {
+                    Ok(path) => {
+                        if path.is_file() {
+                            paths.push(path);
+                        } else if path.is_dir() {
+                            paths.extend(self.get_subfile(path, 0)?);
```

Comment:
So do I need to remove support for input folder? Like this:

```suggestion
                            println!("Input folder is not supported, please input file!");
```

Or keep support for input folders:

```suggestion
                            for entry in path.read_dir()? {
                                let path = entry?.path();
                                if path.is_file() {
                                    paths.push(path);
                                }
                            }
```

Reply:
Great! I think should remove the support for input folder. But maybe you should skip processing this path when it's a dir, because sometimes the `glob` gets to the directory.

I'm sorry, this is my fault.